src/coreclr/jit/gentree.cpp

```diff
@@ -3486,6 +3486,7 @@ unsigned Compiler::gtSetEvalOrder(GenTree* tree)
                 }
 #elif defined(TARGET_ARM64)
                 if ((*((__int64*)&(tree->AsDblCon()->gtDconVal)) == 0) ||
+                    (compCurBB->bbWeight > BB_UNITY_WEIGHT) ||
```

Comment:
The condition should be `&&` with hot block condition.

```c++
if ((*((__int64*)&(tree->AsDblCon()->gtDconVal)) == 0) || ((compCurBB->bbWeight > BB_UNITY_WEIGHT) && emitter::emitIns_valid_imm_for_fmov(tree->AsDblCon()->gtDconVal))
```

Basically, we want to avoid `fmov` only if it is inside a hot block.

Reply:
Our CSE heuristics around floats are most certainly lacking. We may need to tune them better before we can be more aggressive with CSEing cheap trees while avoiding regressions.